i23: Show game creation screen

nickmeinhold opened the pull request:
fixes #23

nickmeinhold committed:
Add a screen and navigate from lobby

- had to update pods to get ios app to build after firebase migration

nickmeinhold committed:
Merge in dev

nickmeinhold committed:
Merge branch 'dev' into nickmeinhold/issue23

nickmeinhold committed:
Change to new navigation and add FAB

- added named routes for NewGameScreen1 and NewGameScreen2
- changed lobby-to-new_game1 navigation to new navigation method
- split up new NewGameScreen1 into separate widgets
- added a FloatingActionButton that navigates to NewGameScreen2
- added a NewGameScreen2 widget with just a Text widget for now

nickmeinhold committed:
Restructure with viewmodel

- added a viewmodel for NewGameScreen1
- moved widgets for NewGameScreen1 into separate files in a new folder
- changed NewGameScreen1 to use a viewmodel, passed to children

nickmeinhold committed:
Use autovalidation to update after submit

nickmeinhold committed:
Merge branch 'dev' into nickmeinhold/issue23

nickmeinhold committed:
Merge dev in

nickmeinhold committed:
Add view model to game service

nickmeinhold committed:
Add a widget test for NewGameScreen1

nickmeinhold committed:
Merge branch 'nickmeinhold/issue23' of github.com:tamari-gray/niira into nickmeinhold/issue23

nickmeinhold committed:
Add more testing

nickmeinhold committed:
Fix test and move vm insde child widgets

- don't forget ol reliable

tamari-gray commented on lib/screens/new_game1/new_game_screen1.dart:
```diff
@@ -0,0 +1,53 @@
+              autovalidate: _autoValidateForm,
+              key: _formKey,
+              child: Column(
+                children: <Widget>[GameNameField(), PasswordField()],
```
ok that does look vedy clean

tamari-gray commented on lib/screens/new_game1/password_field.dart:
```diff
@@ -0,0 +1,36 @@
+            ),
+            suffixIcon: Icon(Icons.visibility)),
+        validator: validators.oneToFifteenChars,
+        onChanged: (val) => setState(() => _vm?.password = val),
```
what happens if vm is null?

nickmeinhold replied:
Good question! I guess it would break, although I'm not sure.  I believe it can never be null when the function runs because it's set in initState... I think maybe the null safety operator was to shut the analyzer up, I don't really remember.  Anyway it's a hangover from having the vm created in the parent widget, now it's in the service I can get rid of this nonsense.  Good pickup, thanks!

tamari-gray replied:
alg!

nickmeinhold replied:
So I was wrong that it will break - if you put something like this into dartpad, it works fine:

```Dart
class Classy {
  String stringy;  
}

void main() {
  Classy c;
  c?.stringy = 'hello';
  
  print(c);
}
```

I don't really get why it doesn't break though so even though the behaviour seems to be what we would want, I don't like it... who wrote this code anyway?  

When I try and make it nice though, it all breaks.  I would like to make the name and password widgets stateless and just get the VM out of the service as needed, but before I even get to stateless, if I just replace the `_vm?` with `context.read<GameService>().newGameViewModel1` the test fails, apparently because navigation didn't occur (presumably because the valid state was never obtained?). 

What would you like to do?  Take the code as is that works but is a bit strange or spend the time figuring out why changing it breaks?  I don't mind but I won't be able to spend any time on it tonight.

nickmeinhold replied:
Actually put the laptop down!  Rest up and get rid of that headache 🤒

tamari-gray replied:
hi @nickmeinhold, i just made a branch and converted them to stateless widgets. not sure why it was breaking for you, but when you get a chance can you please have a look at it and see if it helps at all. happy to merge this as it is too :)

nickmeinhold replied:
Aaaah I think I get it now!  

So the answer to your original question should have been:

If _vm is null in 

`_vm?.password = val`

then

`password = val`

just doesn't get called. 

It's the weird syntax of setters that got us ;-)  everything after the ? is just a function and if the object that the function is called on is null, the function doesn't get called.  If the line had have been: 

`_vm?.setPassword(val)`

I imagine neither of us would have been tricked.  

Well that was a great learning exercise anyway, thanks for the question!

tamari-gray commented on lib/utilities/validators.dart:
```diff
@@ -0,0 +1,3 @@
+final oneToFifteenChars = (String value) => value.isEmpty
```
love that you put this into a utilities folder, makes me want to refactore the create user account screen

nickmeinhold replied:
Thanks!  I liked this pattern too - top level functions in a utilities file and then import with a prefix... you don't even need a class but avoid namespace collisions

tamari-gray commented on test/create_game_section/new_game1/new_game1_widget_test.dart:
```diff
@@ -0,0 +1,73 @@
+      expect((nameField), findsOneWidget);
+      expect((passwordField), findsOneWidget);
+
+      // add invalid data
```
should be // add valid data

nickmeinhold replied:
Legend!

tamari-gray reviewed (changes requested):
great work (as always) thank nick!

i though everything was perfect and i learnt alot from your pr!
was really easy to read through & understand. 

just that "invalid" typo (refer to comment) to fix up then very happy to merge!

sorry if my comments were short, was just feeling lazy as i wrote them 😄

nickmeinhold commented:
Thanks for the comments mate, will make a couple of fixes now and re-request review

nickmeinhold committed:
Fix comment in test

tamari-gray commented on test/create_game_section/new_game1/new_game1_widget_test.dart:
```diff
@@ -57,7 +57,7 @@ void main() {
       expect((passwordField), findsOneWidget);
 
-      // add invalid data
+      // add valid data
```
beautiful haha

tamari-gray reviewed (approved):
thanks for the work nick!
glad you got to eat some 🥩, hopefully have some 🍰 tickets coming 🙏

tamari-gray reviewed (commented):
sorry @nickmeinhold dont merge yet, im just looking at that issue with ?vm

tamari-gray commented on lib/screens/new_game1/password_field.dart:
```diff
@@ -0,0 +1,36 @@
+}
+
+class _PasswordFieldState extends State<PasswordField> {
+  NewGameViewModel1 _vm;
```
forgot to init with context.read game serice 👷‍♂️

tamari-gray reviewed (changes requested):
hi @nickmeinhold, can you please look at my pr into this branch and tell me what you think. Im not sure if im using a viewmodel strucure wrong. but happy to talk whenever today!

tamari-gray committed:
part of 92, convert form feilds to stateless widgets (#101)

* convert to stateless widgets with null check

* meant to be with last commit

Co-authored-by: tamari-gray <tamari-gray>

nickmeinhold commented:
No rush, please enjoy your lunch! 😄

nickmeinhold commented on lib/screens/new_game1/game_name_field.dart:
```diff
@@ -0,0 +1,29 @@
+        ),
+        validator: validators.oneToFifteenChars,
+        onChanged: (val) =>
+            context.read<GameService>()?.newGameViewModel1?.name = val,
```
@tamari-gray you know what I still don't get - is why we need the null checks... 

I get why we needed them previously, when the _vm object was only set in initState, but why now?  The test breaks without them now I think but why is the vm or the game service ever null? (I guess it must be or the test wouldn't break) 

anyway not worth holding up  the PR imo, just wanted to see if you had any thoughts

nickmeinhold replied:
Oh maybe there's an onChanged event in the first build of the widget tree, before everything is set up?

nickmeinhold replied:
I think I figured it out, we're using a mock service without all the things... got a fix a brewing...

nickmeinhold committed:
Fix test that was using mock game serivce

- changed to using real game service in tests
- removed uneccessary null checks from acessing game service in widgets
- moved viewmodel unit test inside game service unit test

tamari-gray commented:
looks beautiful 🙏
